internal/services/serviceprincipals/service_principal_certificate_resource_msgraph.go

```diff
@@ -0,0 +1,164 @@
+			if cred.KeyId != nil && *cred.KeyId == *credential.KeyId {
+				return tf.ImportAsExistsDiag("azuread_service_principal_certificate", id.String())
+			}
+			newCredentials = append(newCredentials, cred)
+		}
+	}
+
+	newCredentials = append(newCredentials, *credential)
+
+	properties := msgraph.ServicePrincipal{
+		ID:             &id.ObjectId,
+		KeyCredentials: &newCredentials,
+	}
+	if _, err := client.Update(ctx, properties); err != nil {
+		return tf.ErrorDiagF(err, "Adding certificate for service principal with object ID %q", id.ObjectId)
+	}
+
+	d.SetId(id.String())
+
+	return servicePrincipalCertificateResourceReadMsGraph(ctx, d, meta)
+}
+
+func servicePrincipalCertificateResourceReadMsGraph(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).ServicePrincipals.MsClient
+
+	id, err := parse.CertificateID(d.Id())
+	if err != nil {
+		return tf.ErrorDiagPathF(err, "id", "Parsing certificate credential with ID %q", d.Id())
+	}
+
+	app, status, err := client.Get(ctx, id.ObjectId)
+	if err != nil {
+		if status == http.StatusNotFound {
+			return tf.ErrorDiagPathF(nil, "service_principal_id", "Service principal with object ID %q was not found", id.ObjectId)
+		}
+		return tf.ErrorDiagPathF(err, "service_principal_id", "Retrieving service principal with object ID %q", id.ObjectId)
+	}
+
+	var credential *msgraph.KeyCredential
+	if app.KeyCredentials != nil {
+		for _, cred := range *app.KeyCredentials {
+			if cred.KeyId != nil && *cred.KeyId == id.KeyId {
+				credential = &cred
+				break
+			}
+		}
+	}
+
+	if credential == nil {
+		log.Printf("[DEBUG] Certificate credential %q (ID %q) was not found - removing from state!", id.KeyId, id.ObjectId)
+		d.SetId("")
+		return nil
+	}
+
+	tf.Set(d, "service_principal_id", id.ObjectId)
+	tf.Set(d, "key_id", id.KeyId)
+
+	keyType := ""
+	if v := credential.Type; v != nil {
+		keyType = *v
+	}
+	tf.Set(d, "type", keyType)
+
+	startDate := ""
+	if v := credential.StartDateTime; v != nil {
+		startDate = v.Format(time.RFC3339)
+	}
+	tf.Set(d, "start_date", startDate)
+
+	endDate := ""
+	if v := credential.EndDateTime; v != nil {
+		endDate = v.Format(time.RFC3339)
+	}
+	tf.Set(d, "end_date", endDate)
+
+	return nil
+}
+
+func servicePrincipalCertificateResourceDeleteMsGraph(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).ServicePrincipals.MsClient
+
+	id, err := parse.CertificateID(d.Id())
+	if err != nil {
+		return tf.ErrorDiagPathF(err, "id", "Parsing certificate credential with ID %q", d.Id())
+	}
+
+	tf.LockByName(servicePrincipalResourceName, id.ObjectId)
+	defer tf.UnlockByName(servicePrincipalResourceName, id.ObjectId)
+
+	app, status, err := client.Get(ctx, id.ObjectId)
+	if err != nil {
+		if status == http.StatusNotFound {
+			log.Printf("[DEBUG] Service Principal with Object ID %q was not found - removing from state!", id.ObjectId)
+			return nil
+		}
```

Comment:
(as elsewhere) this hides a bug within the Read/refresh where we're not removing this from the state - so can we remove this?